### PR TITLE
feat(space): validate warm_start_configs with clear errors (#220)

### LINKED ISSUE #220: Add clear validation errors for invalid warm_start_configs
## Summary

`warm_start_configs` are useful for seeding the first generation with known-good parameter values, but invalid entries are currently easy for a beginner to misunderstand. If a config has a misspelled key, a missing key, or a value outside the search space, the failure should clearly explain what is wrong and how to fix it.

## Why this helps

New users often copy a `param_grid` and then add `PopulationConfig(warm_start_configs=[...])`. A tiny mismatch like `max_depths` instead of `max_depth`, or `max_depth=None` when the search space is `Integer(2, 20)`, should produce a helpful message instead of failing later during population creation.

## Suggested implementation

Start by reading these files:

- `sklearn_genetic/population.py`
  - `_is_dimension_value_valid()` already knows how to check whether one value belongs to a dimension.
  - `random_search_population()` and `smart_search_population()` both call `estimator.space.sample_warm_start(config)`.
- `sklearn_genetic/space/space.py`
  - Look for the `Space` class and its warm-start sampling helper.
- `sklearn_genetic/tests/test_internal_helpers.py`
  - Existing tests already cover warm starts and population initialization.

A good solution would add validation close to the `Space.sample_warm_start(...)` path, so both `random` and `smart` population initializers benefit from the same error message.

The error should ideally include:

- the warm-start config index, for example `warm_start_configs[0]`
- the offending parameter name
- whether the key is unknown, missing, or has an invalid value
- the expected search-space keys

Example message shape:

```text
Invalid warm_start_configs[0] entry for 'max_depth': value None is not valid for Integer(lower=2, upper=20). Expected keys: ['criterion', 'max_depth'].
```

## Tests to add

Add focused tests in `sklearn_genetic/tests/test_internal_helpers.py` or `sklearn_genetic/space/tests/test_space.py` for:

1. a warm-start config with an unknown key, e.g. `{"max_depths": 4}`
2. a warm-start config with a missing key, e.g. only `{"max_depth": 4}` when the space also has `criterion`
3. a warm-start config with an invalid value, e.g. `{"max_depth": 99}` for `Integer(1, 5)`
4. a valid warm-start config still works

## Validation

Run:

```bash
pytest sklearn_genetic/tests/test_internal_helpers.py sklearn_genetic/space/tests/test_space.py
```

## Notes for contributors

This is a good first issue because the behavior is small and well-contained, but it adds real value: users get a clear fix-it message before a long search starts.

### PR DESCRIPTION
**Pull request by mayoka0**
## Summary

Adds up-front validation of `warm_start_configs` so a typo or out-of-range value produces a clear, actionable error instead of failing later during population creation.

## Related Issue

Closes #220

## Type of Change

- [x] New feature (validation + error messages)

## What changed

`Space.sample_warm_start()` now validates each config and raises for:
- a non-dict config
- an unknown/misspelled hyperparameter name — lists the valid names (e.g. `max_depths` → suggests `max_depth`)
- a value outside its dimension — shows the dimension

Examples:
```
warm_start_configs contains unknown hyperparameter(s) ['max_depths'] that are not
in the search space. Valid names are ['criterion', 'max_depth']. (Check for typos,
e.g. 'max_depths' instead of 'max_depth'.)

warm_start_configs value 99 for 'max_depth' is outside its search space
Integer(lower=2, upper=20, distribution='uniform').
```

Both `random_search_population` and `smart_search_population` call `sample_warm_start`, so both initializers benefit. The value check is factored into `space.value_in_dimension` and reused by `population._is_dimension_value_valid` to keep the two in sync (single source of truth).

## Checklist

- [x] Tests pass locally (`pytest sklearn_genetic/`) — space suite (33) + internal-helpers (37) green
- [x] Code is formatted with Black
- [x] New functionality is covered by tests (4 Space-level + 1 population-initializer integration test)
- [ ] Documentation updated if needed — n/a

— Contributed by **Mayoka Labs**

### CODE REVIEW

**Comment by mayoka0** [comment]
Small heads-up — I went back over this one and caught an edge case myself: the value check used plain `isinstance(value, int)` / `(int, float)`, so a warm-start value like `np.int64(5)` or `np.float64(0.5)` (pretty common when people pull values from arrays or earlier results) would get wrongly flagged as "outside the search space."

Just pushed a fix: it now uses `numbers.Integral` / `numbers.Real`, so NumPy scalars pass while strings and out-of-range values are still rejected. Added a test for it, and updated the troubleshooting docs since that section still said invalid configs are "silently skipped" (they raise a clear error now).

Let me know if you'd rather I split the doc tweak out or handle anything differently — happy to adjust. 🙂

**Comment by mayoka0** [comment]
One more small thing — I noticed this PR and the type-hints one (#230) both touched the import lines at the top of `space.py`, which would've caused a needless merge conflict depending on which lands first. Reworked the value check to use the already-imported `np.integer` / `np.floating` instead of pulling in `numbers`, so this PR no longer touches those import lines and the two merge cleanly in any order. Behaviour's identical. 🙂

**Comment by mayoka0** [comment]
@rodrigo-arenas heads-up — CI caught something interesting, and it's a good sign the validation works. `test_expected_ga_schedulers` was seeding `warm_start_configs` with `max_iter=400`, but that test's `param_grid` declares `max_iter` as `Integer(700, 1000)` — so the seed was actually outside the search space. The old code copied it through silently; the new validation (correctly) flags it.

I bumped the seed to `800` (in range) so the test reflects a valid config — its intent is unchanged. CI should be green now. 🙂

**Comment by mayoka0** [comment]
@rodrigo-arenas quick update after watching CI — I scaled this one back a bit, and wanted to flag why.

My first pass also rejected warm-start *values* that fall outside their dimension (issue #220 mentions that case). But CI caught that your existing `test_space_sample_warm_start_fills_missing_parameters` deliberately asserts the opposite — that `sample_warm_start` passes a provided value through as-is, even out of range. So I'd have been changing an intended contract.

I've scoped this PR to the safe, non-breaking part: it rejects a non-dict config and unknown/misspelled keys (the `max_depths` vs `max_depth` case from the issue) and leaves provided values untouched. Reverted the range check and a couple of related bits. Green now, and no existing tests change behaviour.

**Your call on the range part:** if you do want out-of-range values rejected too, I'm happy to add it — but it'd mean updating that passthrough test/contract, so I left that decision to you. 🙂

**Comment by rodrigo-arenas** [comment]
Great work, @mayoka0 — the unknown-key + non-dict validation is clean, well-tested, and CI is green. Nice catches on the NumPy scalar case, the #230 import conflict, and the `max_iter=400` seed in an `Integer(700, 1000)` space.

On your open question: let's add the out-of-range value check too. It fits the "fail early with a clear error" direction of #220/#221/#222, and the scheduler bug you found is exactly what it should catch. A few guardrails:

1. **Only the out-of-range case** — keep missing-key configs working (partial warm-starts that fill by sampling are intended, so don't implement #220's "missing key").
2. **Reuse `population._is_dimension_value_valid`** (e.g. via a shared `Space.value_in_dimension`) so the checks can't drift.
3. Update `test_space_sample_warm_start_fills_missing_parameters` and the troubleshooting docs (currently say values are "used as-is") to match the new contract.

Optional: include the config index (`warm_start_configs[0]`) in the message, like the issue's example.

Thanks again — almost there.

**Comment by mayoka0** [comment]
@rodrigo-arenas done — added the out-of-range value check with your guardrails:
- rejects values outside their dimension; **missing keys still allowed** (filled by sampling)
- added a shared `Space.value_in_dimension` and routed `population._is_dimension_value_valid` through it so they can't drift (NumPy scalars accepted)
- updated `test_space_sample_warm_start_fills_missing_parameters` + the troubleshooting docs (no longer 'used as-is'), and re-set the schedulers seed back into range

Left the config-index in the message out for now since it was optional — happy to add `warm_start_configs[0]` to the wording if you'd like it. 🙂

**Comment by rodrigo-arenas** [comment]
Thanks @mayoka0 merging this
